grimoire/core/styling/palette.py

```diff
@@ -0,0 +1,148 @@
+from grimoire.core.styling.materials.material import Material, MaterialParameters
+
+
+class MaterialRole(Enum):
+    primary_wall = "primary_wall"
+    secondary_wall = "secondary_wall"
+
+    primary_roof = "primary_roof"
+    secondary_roof = "secondary_roof"
+
+    primary_stone = "primary_stone"
+    secondary_stone = "secondary_stone"
+
+    primary_wood = "primary_wood"
+    secondary_wood = "secondary_wood"
+
+    pillar = "pillar"
+
+
+class _ResolutionPriority:
+    order: list[MaterialRole]
+    next_in_line: dict[MaterialRole, MaterialRole]
+
+    def __init__(
+        self, order: list[MaterialRole], next_in_line: dict[MaterialRole, MaterialRole]
+    ):
+        self.order = order
+        self.next_in_line = next_in_line
+
+    def get_next_in_line(self, role: MaterialRole) -> MaterialRole | None:
+        if role in self.next_in_line:
+            return self.next_in_line[role]
+
+        return None
+
+
+class ResolutionPriority(Enum):
+    stone = _ResolutionPriority(
+        order=[
+            MaterialRole.pillar,
+            MaterialRole.primary_wall,
+            MaterialRole.secondary_wall,
+            MaterialRole.primary_roof,
+            MaterialRole.secondary_roof,
+            MaterialRole.primary_stone,
+            MaterialRole.secondary_stone,
+            MaterialRole.primary_wood,
+            MaterialRole.secondary_wood,
+        ],
+        next_in_line={
+            MaterialRole.primary_wall: MaterialRole.primary_stone,
+            MaterialRole.secondary_wall: MaterialRole.secondary_stone,
+            MaterialRole.primary_roof: MaterialRole.primary_stone,
+            MaterialRole.secondary_roof: MaterialRole.secondary_stone,
+            MaterialRole.primary_stone: MaterialRole.primary_wood,
+            MaterialRole.secondary_stone: MaterialRole.secondary_wood,
+            MaterialRole.primary_wood: MaterialRole.primary_stone,
+            MaterialRole.secondary_wood: MaterialRole.secondary_stone,
+            MaterialRole.pillar: MaterialRole.primary_stone,
+        },
+    )
+
+    def order(self) -> list[MaterialRole]:
+        return self.value.order
+
+    def get_next_in_line(self, role: MaterialRole) -> MaterialRole | None:
+        return self.value.get_next_in_line(role)
+
+
+class Palette(Asset):
+    resolution_priority: ResolutionPriority
+    materials: dict[MaterialRole, Material]
+
+    # Colors
+    primary_color: MinecraftColor
+    secondary_color: MinecraftColor
+
+    def find_role(self, block: Block) -> MaterialRole | None:
+        for role in self.resolution_priority.order():
+            if role in self.materials and self.materials[role].has_block(block):
+                return role
+
+        return None
+
+    def find_block_id(
+        self,
+        form: BlockForm,
+        parameters: MaterialParameters,
+        role: MaterialRole,
+    ) -> str | None:
+        checked_roles = set()
+
+        while True:
```

Comment:
Why the loop? Could this be a limited loop instead?

Reply:
Basically, I want this to check roles for the block, going to alternative roles via `resolution_priority.get_next_in_line(role)`. It should return `None` once all alternatives have been exhausted. 

I could construct the list for next_in_line first, but instead I decided to just end the loop when we either run out of alternatives or hit a cycle.

Reply:
I'll have to look over this again, but why not make this a for loop and have the code above be a generator to iterate over?